src/matchmaking_servers.rs

```diff
@@ -0,0 +1,556 @@
+
+            game_description: CStr::from_ptr(raw.m_szGameDescription.as_ptr())
+                .to_string_lossy()
+                .into_owned(),
+            server_name: CStr::from_ptr(raw.m_szServerName.as_ptr())
+                .to_string_lossy()
+                .into_owned(),
+            game_dir: CStr::from_ptr(raw.m_szGameDir.as_ptr())
+                .to_string_lossy()
+                .into_owned(),
+            map: CStr::from_ptr(raw.m_szMap.as_ptr())
+                .to_string_lossy()
+                .into_owned(),
+            tags: CStr::from_ptr(raw.m_szGameTags.as_ptr())
+                .to_string_lossy()
+                .into_owned(),
+
+            last_time_played: Duration::from_secs(raw.m_ulTimeLastPlayed.into()),
+        }
+    }
+}
+
+matchmaking_servers_callback!(
+    Ping;
+    _self;
+    ();
+    responded({}): (info: *const sys::gameserveritem_t => GameServerItem where { GameServerItem::from_ptr(info) }),
+    failed({ free_ping(_self) }): ()
+);
+
+matchmaking_servers_callback!(
+    PlayerDetails;
+    _self;
+    ();
+    add_player({}): (
+        name: *const std::os::raw::c_char => &CStr where { CStr::from_ptr(name) },
+        score: i32 => i32 where {score},
+        time_played: f32 => f32 where {time_played}
+    ),
+    failed({ free_playerdetails(_self) }): (),
+    refresh_complete({ free_playerdetails(_self) }): ()
+);
+
+matchmaking_servers_callback!(
+    ServerRules;
+    _self;
+    ();
+    add_rule({}): (
+        rule: *const std::os::raw::c_char => &CStr where { CStr::from_ptr(rule) },
+        value: *const std::os::raw::c_char => &CStr where { CStr::from_ptr(value) }
+    ),
+    failed({ free_serverrules(_self) }): (),
+    refresh_complete({ free_serverrules(_self) }): ()
+);
+
+matchmaking_servers_callback!(
+    ServerList;
+    _self;
+    (
+        req: Arc<Mutex<ServerListRequest>> where {
+            Arc::new(Mutex::new(ServerListRequest {
+                h_req: ptr::null_mut(),
+                released: false,
+                mms: ptr::null_mut(),
+                real: ptr::null_mut(),
+            }))
+        }
+    );
+    responded({}): (
+        request: sys::HServerListRequest => Arc<Mutex<ServerListRequest>> where { ServerListRequest::get(_self) },
+        server: i32 => i32 where {server}
+    ),
+    failed({}): (
+        request: sys::HServerListRequest => Arc<Mutex<ServerListRequest>> where { ServerListRequest::get(_self) },
+        server: i32 => i32 where {server}
+    ),
+    refresh_complete({}): (
+        request: sys::HServerListRequest => Arc<Mutex<ServerListRequest>> where { ServerListRequest::get(_self) },
+        response: ServerResponse => ServerResponse where {response}
+    )
+);
+
+#[repr(u32)]
+#[derive(Debug, Copy, Clone, Hash, PartialEq, Eq)]
+pub enum ServerResponse {
+    ServerResponded = 0,
+    ServerFailedToRespond = 1,
+    NoServersListedOnMasterServer = 2,
+}
+
+pub struct ServerListRequest {
```

Comment:
Should this include a Drop impl that automatically releases the request?

Reply:
Great question to get a better code understanding! It's impossible to track whether request is still in use in case we also want to also return request structure to manage it from outside a callback, as it's done for C functions which returns the handle. We're basically creating a c++ class layout (that "real" and "virtual" structures) in the heap via boxes. And since the user can decide to, for example, refresh the query from outside, the library itself can't know exactly when it's time to release a request handle and free the above-mentioned structures. So it's important to call `ServerListRequest.release()` at the end of using the request to avoid memory leaks.

Also I've just realized that calling `ServerListRequest.release()` on a refreshing (`SteamAPI_ISteamMatchmakingServers_IsRefreshing`) request can, with a very small chance, cause accessing freed memory. `ServerListRequest.release()` first calls `SteamAPI_ISteamMatchmakingServers_ReleaseRequest` which is necessary (`This request object must be released by calling ReleaseRequest when you're done using it!`) and as the doc says `Releases the asynchronous server list request object and cancels any pending query on it if there's a pending query in progress`. And then calls `free_serverlist` which freeing memory of the whole request. But as said above `ReleaseRequest` cancels only pending query so there is chance that first thread where we called `release` has freed request memory and on other thread which calls `RunCallbacks` steamworks already called "C" callback function (line 20) which accesses those freed structures to call rust function (line 27). The only fix for that i found is not to allow releasing refreshing request, so for example we can return some simple enum from `ServerListRequest.release()` with few variants like `AlreadyReleased` and `Refreshing` inside a `Result`

Taking all this into account I'm not sure if it's worth at all to return Request from function itself. It will make request non-reusable (since you won't be able to refresh query from outside), etc. but on other hand we'll get rid of that ugly ServerListRequest's inside Mutex'es and as far as i think make possible releasing request automatically or while query is refreshing if user wants this. Do You have any thoughts about that?

Steam has made a horrible api.